src/hashtable.c

```diff
@@ -52,12 +59,39 @@ Hashtable* hashtable_new(size_t size, HTCmpFunc cmp_func, HTHashFunc hash_func,
     ht->hash_func = hash_func;
     ht->copy_func = copy_func;
     ht->free_func = free_func;
-    ht->deferred_unsets = NULL;
 
+    SDL_AtomicSet(&ht->cur_operation, HT_OP_NONE);
     assert(ht->hash_func != NULL);
 
     return ht;
 }
+static bool hashtable_try_enter_state(Hashtable *ht, int state, bool mutex) {
+    SDL_atomic_t *ptr = &ht->cur_operation;
+
+    if(!mutex) {
+        SDL_AtomicCAS(ptr, state, HT_OP_NONE);
+    }
+
+    return SDL_AtomicCAS(ptr, HT_OP_NONE, state);
+}
+
+static void hashtable_enter_state(Hashtable *ht, int state, bool mutex) {
+    while(!hashtable_try_enter_state(ht, state, mutex));
+}
+
+static void hashtable_idle_state(Hashtable *ht) {
+    SDL_AtomicSet(&ht->cur_operation, HT_OP_NONE);
+}
+
+void hashtable_lock(Hashtable *ht) {
```

Comment:
It doesn't use them here at all. Notice how they are not static. They are meant to be used by external code with the otherwise thread-unsafe iterators, which may not be the best interface, but at least it abstracts away the whole internal state business.